demos/immersed_fem/immersed_fem.py.rst

```diff
@@ -0,0 +1,115 @@
+
+  Curve Loop( 9) = {1, 2, 3, 4};
+  Curve Loop(10) = {8, 5, 6, 7};
+
+Then, we define two plane surfaces: the rectangle without the disc first, and the disc itself then. ::
+
+  Plane Surface(1) = {9, 10};
+  Plane Surface(2) = {10};
+
+Finally, we group together some edges and define ``Physical`` entities, which
+can be accessed in Firedrake. ::
+
+  Physical Curve("HorEdges", 11) = {1, 3};
+  Physical Curve("VerEdges", 12) = {2, 4};
+  Physical Curve("Circle", 13) = {8, 7, 6, 5};
+  Physical Surface("PunchedDom", 3) = {1};
+  Physical Surface("Disc", 4) = {2};
+
+For simplicity, we have gathered all this commands in the file
+`immersed_domain.geo <immersed_domain.geo>`__. To generate a mesh using this file,
+you can type the following command in the terminal::
+
+    gmsh -2 punched_domain.geo -format msh2
+
+To illustrate how to access all these features within Firedrake,
+we consider the following interface problem. Denoting by
+:math:`\Omega` the filled rectangle and by :math:`D` the disc,
+we seek a function :math:`u\in H^1_0(\Omega)` such that
+
+.. math::
+
+   -\nabla \cdot (\sigma \nabla  u) + u = 5 \quad \textrm{in } \Omega
+
+where :math:`\sigma = 1` in :math:`\Omega \setminus D` and :math:`\sigma = 2`
+in :math:`D`. Since :math:`\sigma` attains different values across :math:`\partial D`,
+we need to prescribe the behavior of :math:`u` across this interface. This is
+implicitely done by imposing :math:`u\in H^1_0(\Omega)`: the function :math:`u` must be continuous
+across :math:`\partial \Omega`. This allows us to employ Lagrangian finite elements
+to approximate :math:`u`. However, we also need to specify the the jump
+of :math:`\sigma \nabla u \cdot \vec{n}` on :math:`\partial D`. This term arises
+naturally in the weak formulation of the problem under consideration. In this demo
+we simply set
+
+.. math::
+
+   [\sigma \nabla u \cdot \vec{n}] = 3 \quad \textrm{on}\ \partial D
+
+.. note::
+   To derive the correct weak formulation it is important to decide whether the normal vector
+   :math:`\vec{n}` points inward or outward. One should also decide the order in which
+   the jump operator is computed. This is irrelevant for the purpose of our demo and may
+   only lead to replacing the sign of the boundary integral in the weak formulation.
+
+The resulting weak formulation reads as follows:
+
+.. math::
+
+   \int_\Omega \sigma \nabla u \cdot \nabla v + uv \,\mathrm{d}\mathbf{x} - \int_{\partial D} 3v \,\mathrm{d}S = \int_{\Omega} 5v \,\mathrm{d}\mathbf{x} \quad \text{for every } v\in H^1_0(\Omega)\,.
+
+The following Firedrake code shows how to solve this variational problem
+using linear Lagrangian finite elements. ::
+
+   from firedrake import *
+   mesh = Mesh('immersed_domain.msh')
+   V = FunctionSpace(mesh, "CG", 1)
+   u = TrialFunction(V)
+   v = TestFunction(V)
+   a = 2*dot(grad(v), grad(u))*dx((4,)) + dot(grad(v), grad(u))*dx((3,)) + v*u*dx
+   L = Constant(5.) * v * dx + Constant(3.)*v*ds((13,))
```

Comment:
Since the jump continuity term is on an interior face, I think this needs to be a `dS` integral.

Note also that you can say `dx(4)` rather than `dx((4, ))`

Reply:
I know use `dS`, thanks